Inline split fixes#12699

splav opened the pull request:
<!-- Please describe your changes on the following line: -->

Fix fragment splitting algorithm to set FIRST/LAST_FRAGMENT_OF_ELEMENT flag correctly, account for border_padding for border_box and set zero border_padding between for new fragments.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #12699 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12760)

<!-- Reviewable:end -->

splav committed:
reduce max_inline_size inside breaking strategy by border_padding

splav committed:
reset border_padding between, account for border_padding in border_box and reset *_FRAGMENT_OF_ELEMENT flag for splitted fragments

splav committed:
add test for splitted span with borders

splav commented:
The thing I'm concerned about is that the test is not fully adequate. The sizes, even in em are font-dependant and the line breakage can happen in different places.
The max_inline_size change makes breakage behave like in Firefox. Honestly, I haven't checked the specs.
And, sorry, I've mixed up commit names, will fix them among other fixes after review.

pcwalton commented on components/layout/fragment.rs:
```diff
@@ -1644,7 +1644,7 @@ impl Fragment {
         };
 
-        let mut remaining_inline_size = max_inline_size;
+        let mut remaining_inline_size = max_inline_size - self.border_padding.inline_start_end();
```
Could you add a comment saying that `self.border_padding.inline_start_end` (presumably) will only be nonzero in the incremental reflow case (I guess?)

This could easily regress.

pcwalton replied:
Disregard previous comment, I was very confused. Sorry about that.

pcwalton commented on components/layout/inline.rs:
```diff
@@ -636,7 +636,23 @@ impl LineBreaker {
+            (Some(mut inline_start_fragment), Some(mut inline_end_fragment)) => {
+                inline_start_fragment.border_padding.inline_end = Au(0);
+                if let Some(ref mut inline_context) = inline_start_fragment.inline_context {
+                    for node in inline_context.nodes.iter_mut() {
```
Could just be `for node in &mut inline_context.nodes {` I think

pcwalton commented on components/layout/inline.rs:
```diff
@@ -636,7 +636,23 @@ impl LineBreaker {
+
+                inline_end_fragment.border_padding.inline_start = Au(0);
+                if let Some(ref mut inline_context) = inline_end_fragment.inline_context {
+                    for node in inline_context.nodes.iter_mut() {
```
Same as above: I believe this could be `for node in &mut inline_context.nodes {`

pcwalton commented:
Just had a couple of review nits to address and then this looks good. Thanks!

pcwalton commented:
Don't worry about the test; as long as it splits in the right place on at least one of our bots it will catch any regressions.

notriddle commented:
Couldn't you use Ahem, the test font?

Ms2ger commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit e68e263 with merge 15ffeba...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2508)

highfive commented:
▶ PASS [expected FAIL] /css21_dev/html4/c5509-ipadn-l-001.htm
    
      ▶ PASS [expected FAIL] /css21_dev/html4/c5509-ipadn-l-003.htm
    
      ▶ PASS [expected FAIL] /css21_dev/html4/inlines-017.htm

notriddle commented:
About your question, you can ask the bot for a particular reviewer using `r?`. For example:

r? @notriddle

splav committed:
update tests expectations

splav commented:
@notriddle I've changed the tests, looking at other test, using Ahem font. Is this version looking better. I've also changed the div width to get the same result.

notriddle commented on tests/wpt/mozilla/tests/css/border_inline_split.html:
```diff
@@ -0,0 +1,20 @@
+<meta content="ahem dom" name="flags">
+<style>
+    div {
+        width: 9em;
```
Make the div 20px Ahem font. That way the em size is consistent everywhere.

notriddle commented:
Other than that, looks awesome! Fix it, then r=me.

splav commented:
Fixed
r=@notriddle

jdm commented:
@bors-servo: r=notriddle
Thanks for jumping on that, @notriddle!

bors-servo commented:
:pushpin: Commit 98df0bf has been approved by `notriddle`

<!-- @bors-servo r=notriddle 98df0bf514d05db006573131c18b353afb12ca2f -->

splav commented:
@jdm argh... I should finally put the bot commands list somewhere in the bookmarks... Thanks.

notriddle commented:
@splav Since I can't delegate, you wouldn't have been able to do it anyway.

splav commented:
@notriddle OK, I see.

bors-servo commented:
:hourglass: Testing commit 98df0bf with merge 7ad120a...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2402)

highfive commented:
▶ FAIL [expected PASS] /css21_dev/html4/c5509-ipadn-l-003.htm
      └   → /css21_dev/html4/c5509-ipadn-l-003.htm 60dc40da063dfa4b8f48965e9324ab4159ec6cca
    /css21_dev/html4/reference/c5502-imrgn-r-003-ref.htm f02ece1ca5437cc3070b1ba6b068ec29f330a545
    Testing 60dc40da063dfa4b8f48965e9324ab4159ec6cca == f02ece1ca5437cc3070b1ba6b068ec29f330a545

notriddle commented:
Platform-specific pass, probably because of Core Fonts. Just put something like this in:

File: `tests/wpt/metadata-css/css21_dev/html4/c5509-ipadn-l-003.htm.ini`
```
[c5509-ipadn-l-003.htm]
  type: reftest
  expected:
    if os == "mac": FAIL
```

jdm commented:
@bors-servo: delegate=notriddle

bors-servo commented:
:v: @notriddle can now approve this pull request

notriddle commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 9c8b3a6 has been approved by `notriddle`

<!-- @bors-servo r=notriddle 9c8b3a6733e8e9e7db44ca6b4dd9e10334c989d3 -->

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #12783

bors-servo commented:
:pushpin: Commit 9c8b3a6 has been approved by `notriddle`

<!-- @bors-servo r=notriddle 9c8b3a6733e8e9e7db44ca6b4dd9e10334c989d3 -->

bors-servo commented:
:hourglass: Testing commit 9c8b3a6 with merge 2420dc4...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2339), [arm64](http://build.servo.org/builders/arm64/builds/2348), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2395), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2523), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2427), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2410), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2485), [windows-dev](http://build.servo.org/builders/windows-dev/builds/293)

emilio commented:
Awesome work here @splav, thanks! :)
